File: web/sc/js/powder.js
/*! Single-Crystal Diffraction · © 2026 Miloš Dubajić · MIT · https://github.com/dubajicmilos/xrays-on-detector */

/**
 * Powder pattern, ported from single_crystal/powder.py.
 *
 * Every reflection out to d min is enumerated and those sharing a d-spacing
 * are summed, so multiplicity is counted by construction rather than looked up
 * from the Laue class. Accidental overlaps -- two unrelated families at the
 * same d -- merge too, which is what a real diffractometer does as well.
 *
 * The Lorentz factor 1 / (sin^2(theta) cos(theta)) applies to all three
 * radiations. The polarisation factor (1 + cos^2(2 theta)) / 2 is x-rays only:
 * neutrons scatter off nuclei with no polarisation dependence, and it is not
 * meaningful for the electron case here either.
 */

import { TWO_PI } from "../../js/physics.js";

export function computePowder(
  structure,
  {
    wavelength = 1.5406,
    radiation = "xray",
    twoThetaMax = 120,
    fwhm = 0.15,
    nPoints = 2000,
    dTol = 1e-5,
  } = {},
) {
  if (!(wavelength > 0)) throw new Error("wavelength must be positive");
  const ttMax = Math.min(twoThetaMax, 179.9);
  // Bragg: lambda = 2 d sin(theta), so the largest reachable 2*theta fixes the
  // smallest d worth enumerating.
  const dMin = wavelength / (2 * Math.sin((ttMax / 2) * (Math.PI / 180)));

  const hkl = structure.hklWithin(dMin);
  if (!hkl.length)
    throw new Error(
      `no reflections with d ≥ ${dMin.toFixed(3)} Å; try a longer wavelength ` +
        "or a smaller 2θ limit",
    );

  const q = new Float64Array(hkl.length / 3);
  const raw = structure.intensity(hkl, radiation, q);

  const rows = [];
  for (let i = 0; i < q.length; i++) {
    const d = TWO_PI / q[i];
    const sinTheta = wavelength / (2 * d);
    if (sinTheta > 1) continue;
    const theta = Math.asin(sinTheta);
    const tt = (2 * theta * 180) / Math.PI;
    if (tt > ttMax) continue;
    let lp = 1 / (Math.sin(theta) ** 2 * Math.cos(theta));
    if (radiation === "xray") lp *= (1 + Math.cos(2 * theta) ** 2) / 2;
    rows.push({
      d,
      tt,
      I: raw[i] * lp,
      hkl: [hkl[3 * i], hkl[3 * i + 1], hkl[3 * i + 2]],
    });
  }
  if (!rows.length) throw new Error("no reflections fall inside the 2θ range");

  // Merge families sharing a d-spacing: sort by d and cut where the relative
  // gap exceeds the tolerance.
  rows.sort((a, b) => b.d - a.d);
  const peaks = [];
  let group = [rows[0]];
  for (let i = 1; i < rows.length; i++) {
    if (Math.abs(rows[i].d - rows[i - 1].d) > dTol * rows[i - 1].d) {
      peaks.push(group);
      group = [];
    }
    group.push(rows[i]);
  }
  peaks.push(group);

  const twoTheta = new Float64Array(peaks.length);
  const intensity = new Float64Array(peaks.length);
  const dOut = new Float64Array(peaks.length);
  const multiplicity = new Int32Array(peaks.length);
  const hklOut = [];
  for (let i = 0; i < peaks.length; i++) {
    const g = peaks[i];
    twoTheta[i] = g.reduce((s, r) => s + r.tt, 0) / g.length;
    intensity[i] = g.reduce((s, r) => s + r.I, 0);
    dOut[i] = g.reduce((s, r) => s + r.d, 0) / g.length;
    multiplicity[i] = g.length;
    // Representative index: the one a crystallographer would write, i.e. the
    // most positive of the family.
    const best = g
      .map((r) => r.hkl)
      .sort((p, r) => p[0] - r[0] || p[1] - r[1] || p[2] - r[2])
      .pop();
    hklOut.push(best);
  }

  let top = 0;
  for (const v of intensity) if (v > top) top = v;
  if (top > 0)
    for (let i = 0; i < intensity.length; i++) intensity[i] *= 100 / top;

  const x = new Float64Array(nPoints);
  const y = new Float64Array(nPoints);
  const sigma = fwhm / (2 * Math.sqrt(2 * Math.LN2));
  for (let i = 0; i < nPoints; i++) x[i] = (ttMax * i) / (nPoints - 1);
  // A Gaussian is dead beyond four sigma, so each peak only touches the points
  // near it; painting the whole trace per peak is what makes this slow.
  const step = ttMax / (nPoints - 1);
  for (let i = 0; i < peaks.length; i++) {
    const I = intensity[i];
    if (I < 1e-6) continue;
    const lo = Math.max(0, Math.floor((twoTheta[i] - 4 * sigma) / step));
    const hi = Math.min(
      nPoints - 1,
      Math.ceil((twoTheta[i] + 4 * sigma) / step),
    );
    for (let j = lo; j <= hi; j++)
      y[j] += I * Math.exp(-0.5 * ((x[j] - twoTheta[i]) / sigma) ** 2);
  }

  return {
    count: peaks.length,
    twoTheta,
    intensity,
    d: dOut,
    multiplicity,
    hkl: hklOut,
    x,
    y,
    wavelength,
    radiation,
  };
}

File: single_crystal/powder.py
"""Powder pattern for the same structure and the same three radiations.

Every reflection out to d_min is enumerated and those sharing a d-spacing are
summed, so the multiplicity is counted by construction rather than looked up
from the Laue class. Accidental overlaps -- two unrelated families at the same
d -- merge too, which is what a real powder diffractometer does as well.

The Lorentz factor 1 / (sin^2(theta) cos(theta)) applies to all three
radiations. The polarisation factor (1 + cos^2(2 theta)) / 2 is X-rays only:
neutrons scatter off nuclei with no polarisation dependence, and it is not
meaningful for the electron case here either.
"""
from __future__ import annotations

import math
from dataclasses import dataclass

import numpy as np

TWO_PI = 2 * math.pi


@dataclass
class PowderPattern:
    two_theta: np.ndarray  # (P,) peak positions, degrees
    intensity: np.ndarray  # (P,) peak intensities, scaled to 100 at the max
    d: np.ndarray  # (P,) d-spacing, Angstrom
    multiplicity: np.ndarray  # (P,) reflections summed into each peak
    hkl: list  # (P,) a representative hkl per peak
    x: np.ndarray  # (n,) trace abscissa, degrees
    y: np.ndarray  # (n,) trace ordinate
    wavelength: float
    radiation: str

    def __len__(self) -> int:
        return len(self.two_theta)


def compute_powder(
    structure,
    wavelength: float = 1.5406,
    radiation: str = "xray",
    two_theta_max: float = 120.0,
    fwhm: float = 0.15,
    n_points: int = 2400,
    d_tol: float = 1e-5,
) -> PowderPattern:
    """Peak list and a Gaussian-broadened trace.

    wavelength in Angstrom; for electrons use scatter.electron_wavelength(kV).
    fwhm is a constant Gaussian width in degrees, which is a stand-in for the
    instrument, not a Caglioti fit.
    """
    if wavelength <= 0:
        raise ValueError("wavelength must be positive")
    tt_max = min(float(two_theta_max), 179.9)
    # Bragg: lambda = 2 d sin(theta), so the largest reachable 2 theta fixes
    # the smallest d worth enumerating.
    d_min = wavelength / (2 * math.sin(math.radians(tt_max / 2)))

    hkl = structure.hkl_within(d_min)
    if len(hkl) == 0:
        raise ValueError(
            f"no reflections with d >= {d_min:.3f} Å; try a shorter wavelength "
            "or a larger 2-theta limit"
        )
    d = structure.d_spacing(hkl)
    inten = structure.intensity(hkl, radiation)

    sin_theta = wavelength / (2 * d)
    ok = sin_theta <= 1.0
    hkl, d, inten, sin_theta = hkl[ok], d[ok], inten[ok], sin_theta[ok]
    theta = np.arcsin(sin_theta)
    two_theta = np.degrees(2 * theta)
    ok = two_theta <= tt_max
    hkl, d, inten, theta, two_theta = (
        hkl[ok], d[ok], inten[ok], theta[ok], two_theta[ok]
    )
    if len(hkl) == 0:
        raise ValueError("no reflections fall inside the 2-theta range")

    lorentz = 1.0 / (np.sin(theta) ** 2 * np.cos(theta))
    if radiation == "xray":
        lorentz = lorentz * (1 + np.cos(2 * theta) ** 2) / 2
    inten = inten * lorentz

    # Merge families sharing a d-spacing. Sorting by d and cutting where the
    # relative gap exceeds the tolerance groups them without an O(N^2) compare.
    order = np.argsort(-d, kind="stable")
    hkl, d, inten, two_theta = hkl[order], d[order], inten[order], two_theta[order]
    cut = np.nonzero(np.abs(np.diff(d)) > d_tol * d[:-1])[0] + 1
    groups = np.split(np.arange(len(d)), cut)

    peaks_tt, peaks_I, peaks_d, peaks_m, peaks_hkl = [], [], [], [], []
    for g in groups:
        if len(g) == 0:
            continue
        peaks_tt.append(float(two_theta[g].mean()))
        peaks_I.append(float(inten[g].sum()))
        peaks_d.append(float(d[g].mean()))
        peaks_m.append(int(len(g)))
        # Representative index: the one a crystallographer would write, i.e.
        # the most positive of the family.
        peaks_hkl.append(tuple(int(v) for v in hkl[g][np.lexsort(hkl[g].T[::-1])][-1]))

    peaks_tt = np.array(peaks_tt)
    peaks_I = np.array(peaks_I)
    scale = peaks_I.max() if peaks_I.size and peaks_I.max() > 0 else 1.0
    peaks_I = 100.0 * peaks_I / scale

    x = np.linspace(0.0, tt_max, int(n_points))
    sigma = fwhm / (2 * math.sqrt(2 * math.log(2)))
    y = np.zeros_like(x)
    # Only peaks with something to contribute are painted; a structure can
    # easily have thousands of extinct ones.
    live = peaks_I > 1e-6
    for tt, I in zip(peaks_tt[live], peaks_I[live]):
        y += I * np.exp(-0.5 * ((x - tt) / sigma) ** 2)

    return PowderPattern(
        two_theta=peaks_tt,
        intensity=peaks_I,
        d=np.array(peaks_d),
        multiplicity=np.array(peaks_m),
        hkl=peaks_hkl,
        x=x,
        y=y,
        wavelength=wavelength,
        radiation=radiation,
    )
